sound/soc/sof/sof-audio.c

```diff
@@ -490,10 +500,47 @@ int sof_machine_register(struct snd_sof_dev *sdev, void *pdata)
 	mach = plat_data->machine;
 	size = sizeof(*plat_data->machine);
 
-	/* register machine driver, pass machine info as pdata */
-	plat_data->pdev_mach =
-		platform_device_register_data(sdev->dev, drv_name,
-					      PLATFORM_DEVID_NONE, mach, size);
+	/* register machine driver */
+#if IS_ENABLED(CONFIG_SND_SOC_SOF_NOCODEC)
+	if (!strcmp(drv_name, "sof-nocodec")) {
+		struct auxiliary_device *auxdev;
+		int ret;
+
+		auxdev = kzalloc(sizeof(*auxdev), GFP_KERNEL);
+		if (!auxdev)
+			return -ENOMEM;
+
+		auxdev->dev.release = sof_nocodec_auxdev_release;
+		auxdev->dev.parent = sdev->dev;
+		auxdev->name = "nocodec";
+
+		ret = auxiliary_device_initialize(auxdev);
+		if (ret < 0) {
+			dev_err(sdev->dev, "failed to initialize nocodec dev %s\n", auxdev->name);
+			kfree(auxdev);
+			return ret;
+		}
+
+		ret = auxiliary_device_add(auxdev);
+		if (ret < 0) {
+			dev_err(sdev->dev, "failed to add nocodec dev %s\n", auxdev->name);
+			/*
+			 * auxdev will be freed when the release callback is invoked through
+			 * put_device()
+			 */
+			auxiliary_device_uninitialize(auxdev);
+			return ret;
+		}
+
+		plat_data->nocodec_mach = auxdev;
+		dev_dbg(sdev->dev, "created machine %s\n", dev_name(&auxdev->dev));
+
+		return 0;
```

Comment:
let's put all this in a function, it's taking disproportionately a lot of space for the functionality, only needed for developers. Something like
```
if (!strcmp(drv_name, "sof-nocodec"))
	return nocodec_machine_register(sdev);
```